### ISSUE TITLE
De/Serialization costs.

### ISSUE THREAD

**Issue by EshG** [dotnet/orleans#6709]
I need the grain return value as HashSet. my question is: when I pass (or return) a HashSet. are there any performance benefits to that instead of passing a simple array and constructing the hashset on the other side?

In other words, when I read that:  

> if the method in the grain interfaces is declared to accept IDictionary but at runtime the sender passes SortedDictionary, the receiver will indeed get SortedDictionary.

that is a really cool feature, but are there any costs associated with such convenience that can be avoided by passing only basic types?

**Comment by ReubenBond** [comment]
You wouldn't gain any performance benefits by converting the `HashSet` into an array just to convert it back to a `'HashSet`. On the wire, the `HashSet` is already serialized in the same way as an array (a count followed by each serialized element).

The cost of the convenience is that type names are serialized alongside types, however, there is an optimization: if you say you're sending a `HashSet<int>` and that` exactly what you send, then the type is serialized as `ExpectedType` (one byte) instead of a full type name. Therefore, it is more optimal to serialize concrete types and not abstract types or interfaces.

**Comment by EshG** [comment]
@ReubenBond Thank you for the detailed explanation. and so now I understand that HS->HS is optimal. 
Now I'm afraid that you only answered half of my question, I'll try to be more precise: 
In the grain, I construct a list of integers. and because  

> the HashSet is already serialized in the same way as an array

 it seems logical that converting the list into HS just to satisfy the desired  return type,  is much more expensive than sending the already constructed list/array. 

In other words, when Orleans serialize the HS, all the HS structure is lost (hash map/buckets) and only the elements in their order and total size is preserved?

**Comment by ReubenBond** [comment]
The `HashSet<>` structure is lost, but that is intentional. HS is not an ordered data structure.

`GetHashCode()`, which `HashSet<>` uses, is not guaranteed to return the same value across multiple executions of the same program and therefore you can get different results on different machines in your cluster. Because of this is would be *incorrect* to preserve precise structure across machine boundaries or program executions. `GetHashCode()` often intentionally returns values that are not stable across executions as a DoS mitigation technique.

**Comment by EshG** [comment]
Thanks for confirming.
